rfcs/0042-config-option.md

```diff
@@ -0,0 +1,191 @@
+---
+feature: config-option
+start-date: 2019-03-10
+author: Silvan Mosberger
+co-authors: (find a buddy later to help our with the RFC)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+A lot of NixOS options exist to specify single settings of configuration files. Along with such options come multiple disadvantages, such as having to synchronize with upstream changes, option bloat and more. An alternative is to provide options for passing the configuration as a string, such as the common `configFile` or `extraConfig` options, but this also comes with a set of disadvantages, including difficulty to override values and bad modularity.
```

Comment:
The advantage of having to "synchronize with upstream changes" is that we can provide a graceful transition or at least a warning at upstream changes. If we don't do that, we effectively push the "synchronize with upstream changes" burden to the user.

Reply:
On the other hand, synchornising with upstream changes while knowing the local context is often easier than in general.

Reply:
Actually, I just thought about this again, and we totally can still do whatever we want to. E.g. if upstream changes setting named `foo` to `bar` without backwards compatibility, we can do this:
```nix
{ config, lib, ... }: with lib;
let
  fixedUpConfig = let
    conversion = {
      foo = "bar";
    };
    in attrs: removeAttrs attrs (attrNames conversion) //
      mapAttrs' (name: value:
        nameValuePair value attrs.${name}
      ) (intersectAttrs attrs conversion);
in
  # ...
```

This can't handle nested removals or renames, but this just illustrates how powerful this is: We can arbitrarily modify the configuration however we need too. With a library of config deprecation and rename functions we could handle backwards compatibility easily fully on our side. Not that I want that, but we could. The `config` approach limits us in no way.

Reply:
I added a section mentioning this